README.md
<!-- badges: start -->
[![Project Status: Suspended - Initial development has started, but there has not yet been a stable, usable release; work has been stopped for the time being but the author(s) intend on resuming work.](https://www.repostatus.org/badges/latest/suspended.svg)](https://www.repostatus.org/#suspended)
[![Code size](https://img.shields.io/github/languages/code-size/sfcheung/semunpack.svg)](https://github.com/sfcheung/semunpack)
[![Last Commit at Main](https://img.shields.io/github/last-commit/sfcheung/semunpack.svg)](https://github.com/sfcheung/semunpack/commits/main)
[![R-CMD-check](https://github.com/sfcheung/semunpack/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/sfcheung/semunpack/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

(Version 0.0.1.9009, updated on 2023-06-22, [release history](https://sfcheung.github.io/semunpack/news/index.html))

# IMPORTANT

Due to potential overlap between this package and
[semhelpinghands](https://sfcheung.github.io/semhelpinghands/), I migrated
functions in this package to `semhelpinghands`. This package
will no longer be updated.

# semunpack: Learn SEM Analysis <img src="man/figures/logo.png" align="right" height="150"/>

This package includes functions for learning and teaching structural equation
modeling (SEM). It "unpacks" things inside an SEM analysis. For more information
on this package, please visit its GitHub page:

https://sfcheung.github.io/semunpack/

# Installation

Stable release versions of this package can be downloaded below:

https://github.com/sfcheung/semunpack/releases

The latest developmental version of this package can be installed by `remotes::install_github`:

```r
remotes::install_github("sfcheung/semunpack")
```

# Issues

If you have any suggestions and found any bugs, please feel
feel to open a [GitHub issue](https://github.com/sfcheung/semunpack/issues). Thanks.


NEWS.md
# semunpack 0.0.1.9000+

- (0.0.1.9000+):
    - Work-in-progress version.
    - Drafted `plot_models_fm()`.
    - Drafted `annotate_matrices()`.
    - Drafted `show_ifi()`(0.0.1.9001)
    - Drafted `show_options()` (0.0.1.9002)
    - Drafted `record_history()` (0.0.1.9003)
- Update GitHub actions. (0.0.1.9006)
- Update the pkgdown site. (0.0.1.9007)
- Change to a light theme. (0.0.1.9007)
- Note that this package will no longer be updated. (0.0.1.9009)

tests/testthat.R
library(testthat)
library(semunpack)

test_check("semunpack")


tests/testthat/test-lav_eq.R
# skip_on_cran()
# skip_if(!interactive(),
#         message = "standardizedSolution_boot_ci not tested if not interactive")

library(testthat)
library(semunpack)

options(width = 132)

library(lavaan)

data(HolzingerSwineford1939)
mod <-
"
visual  =~ x1 + v2*x2 + v2*x3
textual =~ x4 + x5 + x6
speed   =~ x7 + x8 + x9
"
fit <- cfa(model = mod, data = HolzingerSwineford1939)

model_gp <-
"
visual  =~ x1 + c(v2, v2)*x2 + c(v2, v3)*x3
textual =~ x4 + x5 + x6
speed   =~ x7 + c(v8, v8)*x8 + c(v9, v9)*x9
"
fit_gp <- cfa(model = model_gp, data = HolzingerSwineford1939, group = "school",
              meanstructure = TRUE)

fit_hist <- record_history(fit)
fit_gp_hist <- record_history(fit_gp)

# plot(fit_hist, params = c("v2", "textual=~x6"),
#             last_n = 10,
#             orientation = "vertical")

test_that("record_history: With equality constraints", {
    expect_equal(
        unlist(as.data.frame(fit_hist[nrow(fit_hist), -c(1:2)])),
        as.vector(coef(fit)),
        ignore_attr = TRUE,
        tolerance = 1e-5
      )
    expect_equal(
        unlist(as.data.frame(fit_gp_hist[nrow(fit_gp_hist), -c(1:2)])),
        as.vector(coef(fit_gp)),
        ignore_attr = TRUE,
        tolerance = 1e-5
      )
  })



tests/testthat/test-annotate_model.R
library(testthat)
library(semunpack)

library(lavaan)

HS.model <- '
visual  =~ x1 + x2 + x3
textual =~ x4 + x5 + x6
speed   =~ x7 + x8 + x9
'
fit_cfa <- cfa(HS.model,
               data = HolzingerSwineford1939)

out <- annotate_matrices(fit_cfa)
out_check <- structure(list(lambda = structure(c("(visual=~x1 = 1)", "visual=~x2",
"visual=~x3", "0", "0", "0", "0", "0", "0", "0", "0", "0", "(textual=~x4 = 1)",
"textual=~x5", "textual=~x6", "0", "0", "0", "0", "0", "0", "0",
"0", "0", "(speed=~x7 = 1)", "speed=~x8", "speed=~x9"), dim = c(9L,
3L), dimnames = list(c("x1", "x2", "x3", "x4", "x5", "x6", "x7",
"x8", "x9"), c("visual", "textual", "speed"))), theta = structure(c("x1~~x1",
"0", "0", "0", "0", "0", "0", "0", "0", "0", "x2~~x2", "0", "0",
"0", "0", "0", "0", "0", "0", "0", "x3~~x3", "0", "0", "0", "0",
"0", "0", "0", "0", "0", "x4~~x4", "0", "0", "0", "0", "0", "0",
"0", "0", "0", "x5~~x5", "0", "0", "0", "0", "0", "0", "0", "0",
"0", "x6~~x6", "0", "0", "0", "0", "0", "0", "0", "0", "0", "x7~~x7",
"0", "0", "0", "0", "0", "0", "0", "0", "0", "x8~~x8", "0", "0",
"0", "0", "0", "0", "0", "0", "0", "x9~~x9"), dim = c(9L, 9L), dimnames = list(
    c("x1", "x2", "x3", "x4", "x5", "x6", "x7", "x8", "x9"),
    c("x1", "x2", "x3", "x4", "x5", "x6", "x7", "x8", "x9"))),
    psi = structure(c("visual~~visual", "visual~~textual", "visual~~speed",
    "visual~~textual", "textual~~textual", "textual~~speed",
    "visual~~speed", "textual~~speed", "speed~~speed"), dim = c(3L,
    3L), dimnames = list(c("visual", "textual", "speed"), c("visual",
    "textual", "speed")))), class = c("annotate_matrices", "list"
))

test_that("Check output", {
    expect_identical(out, out_check)
  })

R/show_ifi.R
#' @title Visualize How CFI and TLI are Computed
#'
#' @description Show how CFI and TLI are computed using a graph of
#'  model chi-square vs. model degrees of freedom.
#'
#' @details This function receives an output of
#'  [lavaan::lavaan()] or its wrappers (e.g., [lavaan::cfa()]
#'  and [lavaan::sem()]) and illustrates how CFI is computed.
#'
#' @return An output of [ggplot2::ggplot()].
#'
#' @param fit An output of [lavaan::lavaan()] or its wrappers (e.g.,
#'  [lavaan::cfa()] and [lavaan::sem()])
#'
#' @param fit_measures The fit measures to be plotted. Acceptable
#'  values are `"cfi"` and `"tli"`.
#'
#' @param test The type of model chi-square test. It corresponds
#'  to the `test` argument of [lavaan::lavaan()] or its wrappers.
#'  Only `"standard"` is supported for now.
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#' @examples
#'
#'
#' library(lavaan)
#'
#' # From the help page of lavaan::cfa().
#'
#' HS.model <- '
#' visual  =~ x1 + x2 + x3
#' textual =~ x4 + x5 + x6
#' speed   =~ x7 + x8 + x9
#' '
#' fit <- cfa(HS.model, data = HolzingerSwineford1939)
#'
#' # By default show how CFI is computed
#' show_ifi(fit)
#'
#' # Show how TLI is computed
#' show_ifi(fit, fit_measures = "tli")
#'
#' # Wrappers
#' show_cfi(fit)
#' show_tli(fit)
#'
#' @export
#'

show_ifi <- function(fit,
                     fit_measures = c("cfi", "tli"),
                     test = c("standard")) {
    fit_measures <- match.arg(fit_measures)
    test <- match.arg(test)
    fit_baseline <- lavaan::lavTech(fit, what = "baseline.test")
    if (is.null(fit_baseline)) {
        stop(paste0("Baseline model is not in 'fit'.",
                    " Was 'baseline' set to 'FALSE'?"))
      }
    if (!(test %in% names(fit_baseline)))  {
        stop(paste0("Test ", dQuote(test),
                    " is not available in 'fit'."))
      }

    df_model <- unname(lavaan::fitMeasures(fit, fit.measures = "df"))
    chisq_model <- unname(lavaan::fitMeasures(fit, fit.measures = "chisq"))

    fit_baseline1 <- fit_baseline[[test]]
    chisq_base <- fit_baseline1$stat
    df_base <- fit_baseline1$df
    chisq_max <- chisq_base * 5 / 4
    df_max <- df_base * 5 / 4

    chisq_df_diff_base <- chisq_base - df_base
    chisq_df_ratio_base <- chisq_base / df_base

    pbase <- ggplot2::ggplot() +
      ggplot2::ylim(0, chisq_max) +
      ggplot2::xlim(0, df_max) +
      ggplot2::labs(
          # title = "Discrepancy (chisq) vs. Simplicity (df)",
          x = "Complex Model <- Degrees of Freedom (df) -> Simple Model",
          y = "Good absolute fit <- Chi-squared (Chisq) -> Bad absolute fit")

    base_point <-
      ggplot2::geom_point(ggplot2::aes(x = df_base,
                                       y = chisq_base),
                          size = 6,
                          colour = "black")
    base_text <-
      ggplot2::annotate("text",
                vjust = 0,
                hjust = .5,
                x = df_base,
                y = chisq_base + chisq_max / 20,
                colour = "red",
                label = paste("Baseline Model\n(df=",
                              df_base,
                              ", Chisq=",
                              formatC(chisq_base, digits = 3, format = "f"),
                              ")",
                              sep = ""))
    saturated_point <-
      ggplot2::geom_point(ggplot2::aes(x = 0,
                                       y = 0),
                          size = 6,
                          colour = "dark green")
    saturated_text <-
      ggplot2::annotate("text",
                vjust = 0,
                hjust = 0,
                x = 0,
                y = 0 + chisq_max / 20,
                colour = "dark green",
                label = "Saturated Model\n(df=0, Chisq=0)")

    df_model <- round(df_base * .75)
    chisq_model <- chisq_base * .20

    p_out <- pbase +
      base_point +
      # base_text +
      saturated_point +
      saturated_text +
      ggplot2::geom_segment(ggplot2::aes(x = 1,
                                         y = 1,
                                         xend = df_base,
                                         yend = df_base),
                            colour = "red",
                            size = 1) +
      ggplot2::geom_point(ggplot2::aes(x = df_base,
                                       y = df_base),
                          size = 4,
                          colour = "red") +
      ggplot2::annotate("text",
                        vjust = 0,
                        hjust = 0.5,
                        x = df_base,
                        y = chisq_base + chisq_max/ 20,
                        label = paste("Baseline Model\n(df=",
                                      df_base,
                                      ", Chisq=",
                                      formatC(chisq_base, digits = 3,
                                              format = "f"),
                                      ")",
                                      sep = ""),
                        colour = "black") +
      ggplot2::annotate("text",
                        vjust = 1,
                        hjust = 0,
                        x = df_base,
                        y = df_base - chisq_max / 40,
                        label = "Red line: Chisq=df",
                        colour = "red") +
      ggplot2::geom_segment(ggplot2::aes(x = df_model,
                                         y = chisq_model,
                                         xend = df_model,
                                         yend = df_model),
                                         size = 1,
                                         colour = "dark green") +
      ggplot2::geom_segment(ggplot2::aes(x = df_model - 2 * df_max / 20,
                                         y = df_model,
                                         xend = df_model + 2 * df_max / 20,
                                         yend = df_model),
                            size = 1,
                            colour = "dark green") +
      ggplot2::geom_point(ggplot2::aes(x = round(df_base * .75),
                                       y = chisq_base * .20),
                                       size = 7,
                                       colour = "blue") +
      ggplot2::annotate("text",
                        vjust = .5,
                        hjust = 1,
                        x = df_model - df_max / 20,
                        y = chisq_model,
                        label = "Fitted model",
                        colour = "blue")

    # CFI
    if (fit_measures == "cfi") {
        p_out <- p_out +
          ggplot2::annotate("text",
                vjust = 0,
                hjust = 0,
                x = 0,
                y = chisq_df_diff_base + chisq_max / 20,
                label = paste0("Blue line:\n",
                               "(Chisq - df) of Fitted Model =\n",
                               "(Chisq - df) of Baseline Model"),
                color = "blue") +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = .5,
                            x = df_base * .75 * .5,
                            y = chisq_base * .40,
                            size = 8,
                            label = "CFI=B/A") +
          ggplot2::geom_segment(ggplot2::aes(x = 1,
                                            y = chisq_df_diff_base,
                                            xend = df_base,
                                            yend = chisq_base),
                                colour = "blue",
                                size = 1) +
          ggplot2::geom_segment(ggplot2::aes(x = df_model + df_max / 20,
                                            y = df_model + chisq_df_diff_base,
                                            xend = df_model + df_max / 20,
                                            yend = df_model),
                                size = 2,
                                colour = "black",
                                arrow = ggplot2::arrow(type = "closed")) +
          ggplot2::geom_segment(ggplot2::aes(x = df_model,
                                            y = chisq_base - round(df_base * (1 - .75)),
                                            xend = df_model,
                                            yend = chisq_base * .20),
                                size = 2,
                                colour = "black",
                                arrow = ggplot2::arrow(type = "closed")) +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = 0,
                            x = df_model + 2 * df_max / 20,
                            y = chisq_base * .50,
                            label = "A",
                            size = 6) +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = 1,
                            x = round(df_base * .75) - df_max / 20,
                            y = chisq_base * .50,
                            label = "B",
                            size = 6) +
          ggplot2::geom_segment(ggplot2::aes(x = df_model - 2 * df_max / 20,
                                            y = df_model + chisq_df_diff_base,
                                            xend = df_model + 2 * df_max / 20,
                                            yend = df_model + chisq_df_diff_base),
                                size = 1,
                                colour = "dark green") +
          ggplot2::labs(title = "The Computation Of CFI")
      }
    # TLI
    if (fit_measures == "tli") {
        p_out <- p_out +
          ggplot2::annotate("text",
                            vjust = 1,
                            hjust = .5,
                            x = df_base / 2,
                            y = chisq_base,
                            label = paste0("Blue line:\n",
                                           "(Chisq/df) of ",
                                           "Fitted Model = \n",
                                           "(Chisq/df) of ",
                                           "Baseline Model"),
                            colour = "blue") +
          ggplot2::geom_segment(ggplot2::aes(x = 1,
                                             y = chisq_df_ratio_base,
                                             xend = df_base,
                                             yend = chisq_base),
                                colour = "blue",
                                size = 1) +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = .5,
                            x = df_base * .75 * .30,
                            y = chisq_base * .5,
                            size = 8,
                            label = "TLI=D/C") +
          ggplot2::geom_segment(ggplot2::aes(x = df_model - 2 * df_max / 20,
                                             y = df_model * (chisq_df_ratio_base),
                                             xend = df_model + 2 * df_max / 20,
                                             yend = df_model * (chisq_df_ratio_base)),
                                             size = 1,
                                             colour = "dark green") +
          ggplot2::geom_segment(ggplot2::aes(x = df_model + df_max / 20,
                                             y = df_model * (chisq_df_ratio_base),
                                             xend = df_model + df_max / 20,
                                             yend = df_model),
                                size = 2,
                                colour = "black",
                                arrow = ggplot2::arrow(type="closed")) +
          ggplot2::geom_segment(ggplot2::aes(x = df_model,
                                             y = df_model * (chisq_df_ratio_base),
                                             xend = df_model,
                                             yend = chisq_model),
                                size = 2,
                                colour = "black",
                                arrow = ggplot2::arrow(type = "closed")) +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = 0,
                            x = df_model + 2 * df_max / 20,
                            y = chisq_base * .50,
                            label = "C",
                            size = 6) +
          ggplot2::annotate("text",
                            vjust = .5,
                            hjust = 1,
                            x = round(df_base * .75) - df_max / 20,
                            y = chisq_base * .50,
                            label = "D",
                            size = 6) +
          ggplot2::labs(title = "The Computation Of TLI")
      }

    p_out
  }

#' @param ... Arguments to be passed to [show_ifi()].
#'
#' @describeIn show_ifi A wrapper of [show_ifi()] with `fit_measures = "cfi"`.
#' @order 2
#' @export

show_cfi <- function(fit, ...) {
    show_ifi(fit, fit_measures = "cfi", ...)
  }

#' @param ... Arguments to be passed to [show_ifi()].
#'
#' @describeIn show_ifi A wrapper of [show_ifi()] with `fit_measures = "tli"`.
#' @order 3
#' @export

show_tli <- function(fit, ...) {
    show_ifi(fit, fit_measures = "tli", ...)
  }

R/show_options.R
#' @title Show Major Options in an Output of 'lavaan'
#'
#' @description Display the values of major options in a model fitted
#'  by [lavaan::lavaan()] or its wrappers (e.g., [lavaan::sem] or
#'  [lavaan::cfa()]).
#'
#' @details It extracts the values of major options in the output of
#'  [lavaan::lavaan()] or its wrappers (e.g., [lavaan::sem] or
#'  [lavaan::cfa()].
#'
#' It checks the actual values, not the call used. This is useful for
#'  understanding how a prepackaged estimator such as `ML`, `MLM`, and
#'  `MLR` set other options. It supports the following options:
#'
#' - Estimator (`estimator`)
#' - Standard error (`se`)
#' - Model chi-square test(s) (`test`)
#' - Missing data method (`missing`)
#' - Information matrix used for computing standard errors (`information`)
#' - Information matrix used for computing model chi-square (`information`)
#' - Whether the mean structure is included.
#'
#' @return
#' A `show_options`-class object with a print method that format the output.
#'
#' @param fit An output of [lavaan::lavaan()] or its wrappers (e.g.,
#'  [lavaan::cfa()] and [lavaan::sem()])
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#' @examples
#' library(lavaan)
#'
#' # From the help page of lavaan::cfa().
#'
#' HS.model <- '
#' visual  =~ x1 + x2 + x3
#' textual =~ x4 + x5 + x6
#' speed   =~ x7 + x8 + x9
#' '
#' fit <- cfa(HS.model, data = HolzingerSwineford1939)
#'
#' tmp <- show_options(fit)
#' tmp
#'
#' fit <- cfa(HS.model, data = HolzingerSwineford1939, estimator = "MLR")
#' show_options(fit)
#' fit <- cfa(HS.model, data = HolzingerSwineford1939, estimator = "MLM")
#' show_options(fit)
#'
#' @export
#'
#' @order 1

show_options <- function(fit) {
    call0 <- lavaan::lavInspect(fit, "call")
    call0_estimator <- call0$estimator
    if (is.null(call0_estimator)) {
        call0_estimator <- lavaan::lavOptions("estimator")$estimator
        call0_estimator <- paste(call0_estimator, collapse = ", ")
      }
    call0_se <- call0$se
    if (is.null(call0_se)) {
        call0_se <- lavaan::lavOptions("se")$se
      }
    call0_missing <- call0$missing
    if (is.null(call0_missing)) {
        call0_missing <- lavaan::lavOptions("missing")$missing
      }
    call0_test <- call0$test
    if (is.null(call0_test)) {
        call0_test <- lavaan::lavOptions("test")$test
      }
    call0_mimic <- call0$mimic
    if (is.null(call0_mimic)) {
        call0_mimic <- lavaan::lavOptions("mimic")$mimic
      }
    call0_model_type <- call0$model.type
    if (is.null(call0_model_type)) {
        call0_model_type <- lavaan::lavOptions("model.type")$model.type
      }
    call0_information <- call0$information
    if (is.null(call0_information)) {
        call0_information <- lavaan::lavOptions("information")$information
      }
    if (length(call0_information) == 1) {
        call0_information1 <- call0_information2 <- call0_information
      } else {
        call0_information1 <- call0_information[1]
        call0_information2 <- call0_information[2]
      }
    call0_meanstructure <- call0$meanstructure
    if (is.null(call0_meanstructure)) {
        call0_meanstructure <- lavaan::lavOptions("meanstructure")$meanstructure
        call0_meanstructure <- ifelse(is.logical(call0_meanstructure),
                                      ifelse(call0_meanstructure, "Yes", "No"),
                                      call0_meanstructure)
      }

    fit_opt <- fit@Options
    opt_estimator <- fit_opt$estimator
    opt_se <- fit_opt$se
    opt_test <- paste(fit_opt$test, collapse = ", ")
    opt_missing <- fit_opt$missing
    opt_mimic <- fit_opt$mimic
    opt_information1 <- fit_opt$information[1]
    opt_information2 <- fit_opt$information[2]
    opt_meanstructure <- ifelse(fit_opt$meanstructure, "Yes", "No")
    call_arg <- c(call0_estimator,
                  call0_se,
                  call0_test,
                  call0_missing,
                  call0_information1,
                  call0_information2,
                  call0_meanstructure)
    opt_arg <- c(opt_estimator,
                 opt_se,
                 opt_test,
                 opt_missing,
                 opt_information1,
                 opt_information2,
                 opt_meanstructure)
    opt_names <- c("Estimator(s)",
                   "Standard Error",
                   "Model Test Statistic(s)",
                   "Missing Data",
                   "Information Matrix (for SE)",
                   "Information Matrix (for Model Test)",
                   "Mean Structure")
    out <- data.frame(Options = opt_names,
                      Call = call_arg,
                      Actual = opt_arg)
    class(out) <- c("show_options", class(out))
    out
  }

#' @param x The output of [show_options()].
#' @param ... Additional arguments. Ignored.
#' @export
#' @describeIn show_options The print method of the output of [show_options()].
#' @order 2

print.show_options <- function(x, ...) {
    class(x) <- class(x)[-1]
    NextMethod(print, x, quote = FALSE, right = FALSE,
               row.names = FALSE, ...)
  }

R/record_history.R
#' @title Record the Minimization History
#'
#' @description Record the minimization history when a model is fitted
#'  by [lavaan::lavaan()] or its wrappers (e.g., [lavaan::sem()] or
#'  [lavaan::cfa()]).
#'
#' @details It records the minimization history when a model is fitted
#'  by [lavaan::lavaan()] or its wrappers (e.g., [lavaan::sem()] or
#'  [lavaan::cfa()]). The recorded history can then be plotted or
#'  displayed, for visualizing and understanding how the estimates
#'  of free parameters is found.
#'
#'  It will refit the model by the update method of
#'  [lavaan::lavaan-class], setting `se = "none"` and `test =
#'  "standard"` because they have no impact effect on the minimization
#'  process.
#'
#' @return
#' A `fit_history`-class object with a plot method.
#'
#' @param fit An output of [lavaan::lavaan()] or its wrappers (e.g.,
#'  [lavaan::cfa()] and [lavaan::sem()]).
#'
#' @author Shu Fai Cheung <https://orcid.org/0000-0002-9871-9448>
#'
#' @examples
#'
#' # Adapted from the example for CFA in lavaan::cfa().
#' # Using only two of the factors
#' library(lavaan)
#' HS.model <-
#' '
#' visual  =~ x1 + x2 + x3
#' textual =~ x4 + x5 + x6
#' '
#' fit <- cfa(HS.model, data = HolzingerSwineford1939)
#' # Refit the model with the history recorded
#' fit_h <- record_history(fit)
#' fit_h
#' plot(fit_h, params = c("visual=~x2", "visual=~x3",
#'                        "visual~~textual"),
#'             last_n = 10)
#' plot(fit_h, params = c("visual=~x2", "visual=~x3",
#'                        "visual~~textual"),
#'             last_n = 10,
#'             orientation = "vertical")
#' @importFrom lavaan coef
#' @export
#'
#' @order 1

record_history <- function(fit) {
    # call_org <- getCall(fit)
    cap <- utils::capture.output(fit1 <- lavaan::update(fit,
                                 control = list(trace = 1),
                                 se = "none",
                                 test = "standard"))
    k <- lavaan::lavInspect(fit1, "iterations")
    capmat <- out2matrix(out = cap[1:(k + 1)],
                         pnames = names(lavaan::coef(fit1)),
                         fit = fit1)
    attr(capmat, "fit_recorded") <- fit1
    attr(capmat, "original_call") <- stats::getCall(fit)
    attr(capmat, "original_output") <- cap
    class(capmat) <- c("fit_history", class(capmat))
    capmat
  }

#' @noRd

out2matrix <- function(out, pnames, fit) {
    out1 <- lapply(out, gsub, pattern = ":", replacement = " ")
    out1 <- lapply(out1, function(x) as.numeric(
                      scan(text = x,
                          what = "",
                          quiet = TRUE)
                    ))
    out1 <- do.call(rbind, out1)
    out1a <- out1[, 1:2]
    out1b <- lav_par_expand(out1[, -c(1:2)], fit = fit)
    out1 <- cbind(out1a, out1b)
    if (!missing(pnames)) {
        colnames(out1) <- c("iteration", "f", pnames)
      }
    as.data.frame(out1)
  }

#' @noRd

plot_history_param <- function(param,
                               fit_history,
                               last_n = -1,
                               orientation = "vertical",
                               ...) {
    history_matrix_tmp <- fit_history
    if (last_n > 0) {
        history_matrix_tmp <- fit_history[
                                    seq(
                                        nrow(history_matrix_tmp) - last_n,
                                        nrow(history_matrix_tmp)
                                      ), ]
      }
    if (orientation == "vertical") {
        x <- history_matrix_tmp[, param]
        y <- history_matrix_tmp[, "f"]
        xlab <- param
        ylab <- "Discrepancy Function Value"
      } else {
        y <- history_matrix_tmp[, param]
        x <- history_matrix_tmp[, "f"]
        ylab <- param
        xlab <- "Discrepancy Function Value"
      }
    plot_args <- list(x = x,
                      y = y,
                      type = "b",
                      xlab = xlab,
                      ylab = ylab,
                      cex = 2,
                      cex.lab = 1.5,
                      lwd = 2)
    plot_args1 <- utils::modifyList(plot_args,
                                    list(...))
    do.call(plot, plot_args1)
    # plot(x, y,
    #      type = "b",
    #      xlab = xlab,
    #      ylab = ylab,
    #      cex = 2,
    #      cex.lab = 1.5,
    #      lwd = 2,
    #      ...)
  }

#' @param x A `fit_history` class object, the output of [record_history()].
#' @param params The names of parameters to be plotted. Must be the
#'               names of one or more columns in `x`.
#' @param last_n The lass `n` iterations to be plotted. Default is -1,
#'               plotting all iterations.
#' @param orientation The orientation of the plot. Either `"horizontal"`
#'               (the default) or `"vertical"`.
#' @param ... Additional arguments. To be passed to [plot.default()]
#'
#' @export
#' @describeIn record_history The plot method for the output of
#'                            [record_history()].
#' @order 2
#' @importFrom graphics axTicks axis mtext par title
plot.fit_history <- function(x,
                             params,
                             last_n = -1,
                             orientation = c("horizontal", "vertical"),
                             ...) {
    fit_history <- x
    orientation <- match.arg(orientation)
    p <- length(params)
    par_old <- par(no.readonly = TRUE)
    on.exit(par(par_old))
    if (last_n < 0) {
        ln0 <- 1
      } else {
        ln0 <- nrow(fit_history) - last_n + 1
      }
    fit_history0 <- fit_history[seq(ln0, nrow(fit_history)), ]
    if (orientation == "vertical") {
        if (p > 1) {
            xaxt0 <- c("s",
                      rep("s", p - 2),
                      "s")
            yaxt0 <- c("s",
                      rep("n", p - 2),
                      "n")
            mai0 <- c(list(    c(1,  1, .1, .1)),
                      rep(list(c(1, .1, .1, .1)), p - 2),
                      list(    c(1, .1, .1, .1)))
            par(mfrow = c(1, p),
                oma = c(5, 5, 3, 2))
            ylab0 <- c("Discrepancy Function Value",
                      rep("", p - 1))
          } else {
            xaxt0 <- c("s")
            yaxt0 <- c("s")
            mai0 <- c(list(    c(1,  1, .1, .1)))
            par(mfrow = c(1, p),
                oma = c(5, 5, 3, 2))
            ylab0 <- c("Discrepancy Function Value")
          }
        xlim0 <- NULL
        ylim0 <- range(fit_history0$f)
        xlab0 <- params
      } else {
        if (p > 1) {
            xaxt0 <- c("n",
                      rep("n", p - 2),
                      "s")
            yaxt0 <- c("s",
                      rep("s", p - 2),
                      "s")
            mai0 <- c(list(    c(.1, 5.5, .1, 2)),
                      rep(list(c(.1, 5.5, .1, 2)), p - 2),
                      list(    c(.1, 5.5, .1, 2)))
            par(mfrow = c(p, 1),
                oma = c(5, 1, 3, 0.5))
            xlab0 <- c(rep("", p - 1),
                      "Discrepancy Function Value")
          } else {
            xaxt0 <- c("s")
            yaxt0 <- c("s")
            mai0 <- c(list(    c(.1, 5.5, .1, 2)))
            par(mfrow = c(p, 1),
                oma = c(5, 1, 3, 0.5))
            xlab0 <- c("Discrepancy Function Value")
          }
        xlim0 <- rev(range(fit_history0$f))
        ylim0 <- NULL
        ylab0 <- params
      }
    for (i in seq_len(p)) {
        par(mar = mai0[[i]])
        plot_history_param(param = params[i],
                           fit_history = fit_history0,
                           last_n = -1,
                           orientation = orientation,
                          #  xaxt = xaxt0[i],
                          #  yaxt = yaxt0[i],
                           frame.plot = TRUE,
                           axes = FALSE,
                           xlim = xlim0,
                           ylim = ylim0,
                           xlab = xlab0[i],
                           ylab = ylab0[i],
                           ...)
        if (orientation == "vertical") {
            if (i == 1) {
                title("Minimization History",
                      line = 1,
                      cex.main = 1.5,
                      outer = TRUE)
                tmp <- formatC(axTicks(2), 4, format = "f")
                axis(2, at = axTicks(2), labels = tmp, las = 1)
                mtext(ylab0[i],
                      side = 2,
                      line = 4,
                      cex = .9)
              }
            tmp <- formatC(axTicks(1), 3, format = "f")
            axis(1, at = axTicks(1), labels = tmp, las = 3)
            mtext(xlab0[i],
                  side = 1,
                  line = 3,
                  cex = .9)
          } else {
            tmp <- formatC(axTicks(2), 2, format = "f")
            axis(2, at = axTicks(2), labels = tmp, las = 1)
            if (i == 1) {
                mtext("Minimization History",
                      side = 3,
                      line = 1,
                      cex = 1.5)
              }
            if (i == p) {
                tmp <- formatC(axTicks(1), 4, format = "f")
                axis(1, at = axTicks(1), labels = tmp)
                mtext(xlab0[i],
                      side = 1,
                      line = 3,
                      cex = .9)
              }
          }
      }
    invisible()
  }

#' @param n_iterations The number of iterations to print. Default is 10,
#'                     printing the first 10 iterations (or all iterations, if
#'                     the number of iterations is less than 10).
#' @param digits The number of digits to be displayed. Default is 3.
#' @param ... Optional arguments. To be passed to the print method of data frame.
#' @export
#' @describeIn record_history The print method for the output of
#'                            [record_history()].
#' @order 3

print.fit_history <- function(x, n_iterations = 10, digits = 3, ...) {
  x1 <- as.data.frame(x)
  n_iterations <- min(nrow(x1), n_iterations)
  cat("Original call:\n")
  print(attr(x, "original_call"))
  cat("The number of iterations:", nrow(x1), "\n")
  if (n_iterations == nrow(x)) {
      cat("The minimization history for all iterations:\n")
    } else {
      cat("The minimization history for the first", n_iterations,
          "iterations:\n")
    }
  cnames <- colnames(x)
  out <- x1[seq_len(n_iterations), -1]
  out <- as.data.frame(lapply(out, formatC, digits = digits, format = "f"),
                       check.names = FALSE)
  out1 <- cbind(iteration = x1[seq_len(n_iterations), 1], out)
  colnames(out1) <- cnames
  print(out1, ...)
  #fit_recorded
  invisible(x)
}